### PR TITLE
Added CommandButtonSingleClickRenderer

### LINKED ISSUE #760: Add single click renderer for CommandButton
See https://github.com/primefaces/primefaces/issues/2521

### PR DESCRIPTION
**Pull request by jepsar**
Fix https://github.com/primefaces-extensions/primefaces-extensions.github.com/issues/760

See also https://github.com/primefaces-extensions/showcase/pull/23

### CODE REVIEW

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        super.encodeEnd(context, component);
+    }
+
+    protected boolean isConfirmation(final CommandButton button) {
+        return button.getStyleClass() != null
+               && Stream.of(CLASS_CONFIRM).anyMatch(s -> s.contains(button.getStyleClass()));
+    }
+
+    protected String toggle(final String widgetVar, final boolean enabled) {
+        return String.format("PF('%s').%sable();", widgetVar, enabled ? "en" : "dis");
```
Comment (Rapster) [inline src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: That line put a smile on my face, this is original! Though, it's `able()` and not `sable()` 😉

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            String widgetVar = button.resolveWidgetVar(context);
+            button.setOnclick(prefix(button.getOnclick(), toggle(widgetVar, false)));
+            button.setOncomplete(prefix(button.getOncomplete(), toggle(widgetVar, true)));
+        }
+        super.encodeEnd(context, component);
+    }
+
+    protected boolean isConfirmation(final CommandButton button) {
+        return button.getStyleClass() != null
+               && Stream.of(CLASS_CONFIRM).anyMatch(s -> s.contains(button.getStyleClass()));
```
Comment (Rapster) [inline src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: I don't think the condition is right, I think you mean `button.getStyleClass().contains(s)`. Being said, instead of creating a stream everytime, I would have write 2 conditions instead:

```java
protected boolean isConfirmation(final CommandButton button) {
        String styleClass = button.getStyleClass();
        return styleClass != null
               && (styleClass.contains("ui-confirmdialog-yes") 
               || styleClass.contains("ui-confirmdialog-no"));
}
```

Why is this necessary btw?

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @author Jasper de Vries &lt;jepsar@gmail.com&gt;
+ * @since 8.0
+ */
+public class CommandButtonSingleClickRenderer extends CommandButtonRenderer {
+
+    protected static final String[] CLASS_CONFIRM = {"ui-confirmdialog-yes", "ui-confirmdialog-no"};
+
+    @Override
+    public void encodeEnd(FacesContext context, UIComponent component) throws IOException {
```
Comment (Rapster) [inline src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Better to override `encodeMarkup` instead no?

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 1 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: 😄 the `%s` is the placeholder, so it's correct.

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 2 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Else the confirm dialog buttons will be disabled on click. You are right about the condition being inverted. I'll change it to what you suggested.

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 3 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Why? I don't think it'll make much of a difference.

Comment (Rapster) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: It doesn't indeed, but `encodeMarkup` is meant for HTML stuff

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 1 above)
Comment (Rapster) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Oh yeah indeed! My bad!

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 3 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: We are modifying attributes and don't really render anything.

Comment (Rapster) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: exactly, HTML attributes

Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Alright, done. Also saved one line 😄

**Review by melloware — APPROVED** [review]
This is nice and clean!

**Review by Rapster — APPROVED** [review]

**Comment by jepsar** [comment]
@melloware can you have a look at the showcase as well please?

https://github.com/primefaces-extensions/showcase/pull/23

**Comment by Rapster** [comment]
But we'll still have the same problem with non ajax command no? Let's say I have a button which validate a form and does a redirect, if form not correctly filled I'll have a validation error and my button will remain disable (since onComplete not called)

**Comment by melloware** [comment]
Maybe we need the enable logic onError too? to renable the button?

**Comment by jepsar** [comment]
Alright, let me have a look at that.

**Comment by jepsar** [comment]
@Rapster I'd suggest to add an Ajax check and only inject the JS for buttons using Ajax. I don't see how we can solve enabling the button for non-Ajax requests, other than having the user to implement something to enable the button again or adding a time out.

@melloware `oncomplete` is also called on error I believe.

**Comment by melloware** [comment]
Agreed for now we should only add this to Ajax buttons.

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            return false;
+        }
+        return button.isAjax();
+    }
+
+    protected boolean isConfirmation(final CommandButton button) {
+        String styleClass = button.getStyleClass();
+        return styleClass != null
+               && (styleClass.contains("ui-confirmdialog-yes")
+                   || styleClass.contains("ui-confirmdialog-no"));
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: What about just one check for styleClass.contains("ui-confirmdialog") ?

**Review by melloware — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 4 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: I considered that.. bit cleaner that way, right?

**Review by melloware — APPROVED** [review]

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    protected void encodeMarkup(FacesContext context, CommandButton button) throws IOException {
+        if (isToggleable(button)) {
+            String widgetVar = button.resolveWidgetVar(context);
+            button.setOnclick(prefix(button.getOnclick(), toggle(widgetVar, false)));
+            button.setOncomplete(prefix(button.getOncomplete(), toggle(widgetVar, true)));
+        }
+        super.encodeMarkup(context, button);
+    }
+
+    protected boolean isToggleable(final CommandButton button) {
```
Comment (Rapster) [inline src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: I don't understand what you mean by `isToggleable`? Also, you can simplify your code with:
```java
return button.isRendered() && !button.isDisabled() && !isConfirmation(button) && button.isAjax()
```

Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: `isToggleable` will tell you if we can toggle enabled / disabled on it. I moved it to a method to prevent an if with the mother of all conditions. I think this is more readable.

What name would be better? needsListeners? disabledStateShouldBeToggled?

Comment (melloware) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: @rapster if you have a better name we are all ears!

Comment (Rapster) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: I don't know @melloware, I only asked to understand but maybe something like `isEligible`/`isEligibleToSingleClick`

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        return button.isAjax();
+    }
+
+    protected boolean isConfirmation(final CommandButton button) {
+        String styleClass = button.getStyleClass();
+        return styleClass != null && styleClass.contains("ui-confirmdialog");
+    }
+
+    protected String toggle(final String widgetVar, final boolean enabled) {
+        return String.format("PF('%s').%sable();", widgetVar, enabled ? "en" : "dis");
```
Comment (melloware) [inline src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: So what abou the scenario in which the Button that was clicked updates the Button so it is no longer rendered.  When the onComplete fires won't the widget not exist to call Enable on?  So I think your original script may be more correct like..
```javascript
var w = PF('%s');if (w){w.%sable();}
```

**Review by melloware — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 6 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Yeah.. I was thinking of why I had those null checks there.. just to be safe, or was I running into issues. Now you are pointing this out I think the latter. I'll change this tonight (CET).

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 5 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: `isEligible` it is!

File: src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java (hunk 6 above)
Comment (jepsar) [reply src/main/java/org/primefaces/extensions/renderer/CommandButtonSingleClickRenderer.java]: Indeed, this check is required (see also update of the showcase)

**Review by melloware — APPROVED** [review]
This was great teamwork!

**Comment by jepsar** [comment]
I think we are good to merge, right? @Rapster ?
